### PR TITLE
[FEATURE] Show projection bounds in projection widget

### PR DESCRIPTION
**Pull request by NathanW2**
## Description
Adds a new bounds view to the projection select dialog.  UI still needs some work but functionally it's there. 

Data is sourced from EPSG.org Postgres dumps. scripts/bounds_sync.py will generate the epsg_areas.csv file that is checked into git anytime it needs to change (maybe once a year or so).

Bounds are synced into the srs.db via syncDatabase at build time.

Image for a demo:

image

I would like to rework the whole projection selection UI but that can be done after merging this.

**NOTE:** I have added a new shapefile to power the map to resources/data which needs to be shipped, however it's quite large at the moment so keen to see if we can get that down but still have a good context for the map.  Open to ideas for this. (Needs to be disconnected because we can't assume the user has net access here)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/gui/qgsprojectionselectiontreewidget.cpp [hunk 1]
```diff
@@ -39,6 +42,16 @@ QgsProjectionSelectionTreeWidget::QgsProjectionSelectionTreeWidget( QWidget *par
...
   setupUi( this );
+  mPreviewBand = new QgsRubberBand( mAreaCanvas, QgsWkbTypes::PolygonGeometry );
+  mPreviewBand->setWidth( 4 );
+
+  QgsCoordinateReferenceSystem srs( 4326, QgsCoordinateReferenceSystem::EpsgCrsId );
+  mAreaCanvas->setDestinationCrs( srs );
+
+  QList<QgsMapLayer *> layers;
+  QString layerPath = QgsApplication::pkgDataPath() + QStringLiteral( "/resources/data/map.shp" );
+  layers << new QgsVectorLayer( layerPath );
```
Comment (nyalldawson) [inline src/gui/qgsprojectionselectiontreewidget.cpp]: layers is leaked here - you'll need to qDeleteAll them the destructor

**Review by nyalldawson — COMMENTED** [review]
Very nice! A couple of suggestions/questions:

1. Do you think the display of the canvas should be controllable? I suspect plugins may be using this widget and would like a way to disable this.

2. I'd add a QgsMapToolPan in here and set the preview canvas to use that tool, so that the map can be zoomed/panned.

3. Unfortunately we've got no way of doing this in QGIS (!!!!!!! please someone sponsor this !!!!!!) but can you run the world map through http://mapshaper.org/ and simplify it to say 5% using Visvalingam weighted?  

4. rename resources/data/map.* to resources/data/world_map.*

5. Shapefile FTW!!! Take that geopackage, you pretentious young upstart.

**Comment by nyalldawson** [comment]
Oh, one last suggestion - the canvas isn't correctly vertically aligned and should take up the full vertical space of the group box.

Oh, and another last thing :P how about a different colored rubber band showing the current canvas extent too? Rip the code from QgsMapCanvasDockWidget::updateExtentRect

**Comment by nyalldawson** [comment]
> Oh, and another last thing :P how about a different colored rubber band showing the current canvas extent too? Rip the code from QgsMapCanvasDockWidget::updateExtentRect

Sorry, getting carried away by the awesomeness of this... but maybe the extent and a marker for the canvas center. I'm just thinking if I'm looking at a map of Burrombuttock and trying to work out what zone I should be using, the map extent probably won't even be visible at the scale of australia, whereas a nice marker for the map center would immediately show me if zone 55 is the correct choice or not...

**Comment by NathanW2** [comment]
Thanks for the feedback.  I'll knock it off tonight.  Keen to get this into a merge state and merged before cut off if no one objects

**Comment by nirvn** [comment]
@NathanW2 , brilliant.

I'm trying to figure out a better UI placement. What about placing the map area next to selected CRS lineedit / proj+ text area:
screenshot from 2017-10-13 12-04-19

The map area would be smaller, but IMHO would be better sized in this context.

**Comment by nirvn** [comment]
@NathanW2 , agreed, we could (should?) get rid of this dual tree view situation. Maybe an editable combobox of recent CRS which would also act as filter?

**Comment by nyalldawson** [comment]
>@NathanW2 , agreed, we could (should?) get rid of this dual tree view situation. Maybe an editable combobox of recent CRS which would also act as filter?

That's a strong -1 from me - in my mapping tasks I've easily got 10-15 regularly used projections which I flip between many many times a day. I'd hate to lose this list!

**Comment by nirvn** [comment]
@nyalldawson , valid point -- putting this list inside a combobox would stil be a -1 for you?

**Comment by NathanW2** [comment]
I'll just move the map to the bottom and leave everything else for now.
One thing at a time and I'm not 100% set on how to manage this UI yet.

**Comment by nirvn** [comment]
@NathanW2 , right, moving the map to the bottom doesn't require the two treeviews to go away.

**Comment by NathanW2** [comment]
image

@nirvn @nyalldawson  canvas extent shown in purple for now.

**Comment by nirvn** [comment]
@NathanW2 , I love it, this needs to be in master now! ;)

**Comment by NathanW2** [comment]
I'll just merge it and let Nyall fix all the bugs ;)

On Sun, Oct 15, 2017 at 11:28 PM, Mathieu Pellerin <notifications@github.com

**Comment by NathanW2** [comment]
@nyalldawson @nirvn center marker added

image

**Comment by nyalldawson** [comment]
> @nyalldawson @nirvn center marker added

Awesome. Now I'll never have to manually lookup what zone Humpybong falls into ever again...

File: src/gui/qgsprojectionselectiontreewidget.cpp [hunk 2]
```diff
@@ -31,13 +36,38 @@ QgsProjectionSelectionTreeWidget::QgsProjectionSelectionTreeWidget( QWidget *par
...
   connect( lstRecent, &QTreeWidget::currentItemChanged, this, &QgsProjectionSelectionTreeWidget::lstRecent_currentItemChanged );
   connect( cbxHideDeprecated, &QCheckBox::stateChanged, this, &QgsProjectionSelectionTreeWidget::cbxHideDeprecated_stateChanged );
   connect( leSearch, &QgsFilterLineEdit::textChanged, this, &QgsProjectionSelectionTreeWidget::leSearch_textChanged );
 
+  mPreviewBand = new QgsRubberBand( mAreaCanvas, QgsWkbTypes::PolygonGeometry );
+  mPreviewBand->setWidth( 4 );
+
+  mPreviewBand2 = new QgsRubberBand( mAreaCanvas, QgsWkbTypes::PolygonGeometry );
+  mPreviewBand2->setWidth( 4 );
+  QColor rectColor =  QColor::fromRgba( qRgba( 185, 84, 210, 60 ) );
```
Comment (nyalldawson) [inline src/gui/qgsprojectionselectiontreewidget.cpp]: This is could be optimised:

   rectColor =  QColor( 185, 84, 210, 60 );

File: src/app/qgsprojectproperties.cpp [hunk 3]
```diff
@@ -154,6 +154,28 @@ QgsProjectProperties::QgsProjectProperties( QgsMapCanvas *mapCanvas, QWidget *pa
   updateGuiForMapUnits( QgsProject::instance()->crs().mapUnits() );
   projectionSelector->setCrs( QgsProject::instance()->crs() );
 
+  QPolygonF mainCanvasPoly = mapCanvas->mapSettings().visiblePolygon();
+  QgsGeometry g = QgsGeometry::fromQPolygonF( mainCanvasPoly );
+  // close polygon
+  mainCanvasPoly << mainCanvasPoly.at( 0 );
+  if ( mapCanvas->mapSettings().destinationCrs() !=
```
Comment (nyalldawson) [inline src/app/qgsprojectproperties.cpp]: Probably safest to use QgsProject::instance()->crs() here, although generally it should be the same as the canvas crs

File: src/core/qgscoordinatereferencesystem.cpp [hunk 4]
```diff
@@ -2103,6 +2110,58 @@ int QgsCoordinateReferenceSystem::syncDatabase()
...
+  {
+    QString line = lines.readLine();
+    if ( line.isNull() )
+      break;
+
+    if ( line.startsWith( '#' ) )
+    {
+      continue;
+    }
+    QStringList data = line.split( "," );
```
Comment (nyalldawson) [inline src/core/qgscoordinatereferencesystem.cpp]: "," -> ','

Comment (nyalldawson) [inline src/core/qgscoordinatereferencesystem.cpp]: const QStringList ...

File: src/core/qgscoordinatereferencesystem.cpp [hunk 5]
```diff
@@ -2103,6 +2110,58 @@ int QgsCoordinateReferenceSystem::syncDatabase()
...
+    if ( data[0] == "None" )
```
Comment (nyalldawson) [inline src/core/qgscoordinatereferencesystem.cpp]: QStringLiteral("None")

File: src/gui/qgsprojectionselectiontreewidget.cpp [hunk 6]
```diff
@@ -31,13 +36,38 @@ QgsProjectionSelectionTreeWidget::QgsProjectionSelectionTreeWidget( QWidget *par
...
+
+  QgsCoordinateReferenceSystem srs( 4326, QgsCoordinateReferenceSystem::EpsgCrsId );
+  mAreaCanvas->setDestinationCrs( srs );
+
+  QString layerPath = QgsApplication::pkgDataPath() + QStringLiteral( "/resources/data/world_map.shp" );
+  mLayers << new QgsVectorLayer( layerPath );
+  mAreaCanvas->setLayers( mLayers );
+  mAreaCanvas->setMapTool( new QgsMapToolPan( mAreaCanvas ) );
+
+  mAreaCanvas->setVisible( mAreaCanvas );
```
Comment (nyalldawson) [inline src/gui/qgsprojectionselectiontreewidget.cpp]: What's this line? SHould it be mShowMap ?

File: src/gui/qgsprojectionselectiontreewidget.h [hunk 7]
```diff
@@ -73,6 +82,14 @@ class GUI_EXPORT QgsProjectionSelectionTreeWidget : public QWidget, private Ui::
     bool showNoProjection() const;
 
     /**
+     * Returns whether the "no/invalid" projection option is shown. If this
```
Comment (nyalldawson) [inline src/gui/qgsprojectionselectiontreewidget.h]: Fix doxygen

File: src/gui/qgsprojectionselectiontreewidget.h [hunk 8]
```diff
@@ -90,6 +107,20 @@ class GUI_EXPORT QgsProjectionSelectionTreeWidget : public QWidget, private Ui::
     void setCrs( const QgsCoordinateReferenceSystem &crs );
 
     /**
+     * Sets the "preview" rectangle for the bounds overview map.
+     * \since QGIS 3.0
+     * \see previewRect()
+     */
+    void setPreviewRect( QgsRectangle rect );
```
Comment (nyalldawson) [inline src/gui/qgsprojectionselectiontreewidget.h]: const QgsRectangle&

File: src/gui/qgsprojectionselectiontreewidget.h [hunk 9]
```diff
@@ -90,6 +107,20 @@ class GUI_EXPORT QgsProjectionSelectionTreeWidget : public QWidget, private Ui::
     void setCrs( const QgsCoordinateReferenceSystem &crs );
 
     /**
+     * Sets the "preview" rectangle for the bounds overview map.
```
Comment (nyalldawson) [inline src/gui/qgsprojectionselectiontreewidget.h]: should this be "initial" preview rectangle?

File: src/gui/qgsprojectionselectiontreewidget.h [hunk 10]
```diff
@@ -90,6 +107,20 @@ class GUI_EXPORT QgsProjectionSelectionTreeWidget : public QWidget, private Ui::
...
+     * \see previewRect()
+     */
+    void setPreviewRect( QgsRectangle rect );
+
+    /**
+     * The "preview" rectangle for the bounds overview map.
+     * \since QGIS 3.0
+     * \see previewRect()
+     */
+    QgsRectangle previewRect();
```
Comment (nyalldawson) [inline src/gui/qgsprojectionselectiontreewidget.h]: const

File: src/gui/qgsprojectionselectiontreewidget.cpp [hunk 11]
```diff
@@ -957,6 +1017,67 @@ long QgsProjectionSelectionTreeWidget::getLargestCrsIdMatch( const QString &sql
...
+{
+  sqlite3      *database = nullptr;
+  const char   *tail = nullptr;
+  sqlite3_stmt *stmt = nullptr;
+
+  QTreeWidgetItem *lvi = lstCoordinateSystems->currentItem();
+  if ( !lvi || lvi->text( QgisCrsIdColumn ).isEmpty() )
+    return;
+
+  int result = sqlite3_open_v2( mSrsDatabaseFileName.toUtf8().data(), &database, SQLITE_OPEN_READONLY, nullptr );
```
Comment (nyalldawson) [inline src/gui/qgsprojectionselectiontreewidget.cpp]: I don't think it should block this PR, but it'd be good to move this bounds retrieval to QgsCoordinateReferenceSystem itself. It'd be quite handy in custom scripts i'd imagine!

**Comment by NathanW2** [comment]
@nyalldawson many thanks for the reviews.

**Comment by rduivenvoorde** [comment]
Cool Nathan!

+1 for using a GeoPackage (which can then maybe used for 
- this little world map
- the extents for the different epsg's
Maybe merge these then with the different sqlite files we have? Adding the srs database in it maybe?

Feature request ( for later :-) ): clicking in the map filters the crs's then spatially, so I can see in one click which crs I have to take for fiji for example.

Great Work going on here \o/

**Comment by gioman** [comment]
> Unfortunately we've got no way of doing this in QGIS (!!!!!!! please someone sponsor this !!!!!!) but can you run the world map through http://mapshaper.org/ and simplify it to say 5% using Visvalingam weighted?

Hi Nyall,
we have v.generalize in the Processing toolbox, that works pretty well.

**Comment by nyalldawson** [comment]
> we have v.generalize in the Processing toolbox, that works pretty well.

Give mapshaper a try. Visvalingam simplification is far far more cartographically useful then the methods available in the grass generalization tools.

**Comment by gioman** [comment]
> Give mapshaper a try. Visvalingam simplification is far far more cartographically useful then the methods available in the grass generalization tools.

ok, will try.

**Comment by wonder-sk** [comment]
It looks like we have Visvalingam simplification in QGIS for some time already: https://github.com/qgis/QGIS/commit/7e1d142fe
... although for now that's used just for on-the-fly simplification

**Comment by gioman** [comment]
> ... although for now that's used just for on-the-fly simplification

does it keep correct topology relations between adjacent polygons?

**Comment by nyalldawson** [comment]
> does it keep correct topology relations between adjacent polygons?

**Comment by nyalldawson** [comment]
> does it keep correct topology relations between adjacent polygons?

Nope - and that's what's missing

**Comment by rduivenvoorde** [comment]
Nice (non topological, though) demo and info: https://bost.ocks.org/mike/simplify/

@nyalldawson is this possible to do it? Is that what mapshaper is doing then?
Because I think it is only possible IF you build a topological dataset first?
(Mike Bostock even thanks mapshaper :-) )

**Pull request merged**